### DISCUSSION TITLE
Receive a "Cannot access Disposed Object" exception when accessing HttpClient from an AWS Lambda Function

### DISCUSSION THREAD

**Discussion by johnt84** [aws/aws-lambda-dotnet#1415] [Q&A]
Receive a "Cannot access Disposed Object" exception when accessing HttpClient from an AWS Lambda Function when the AWS Lambda function is called a 2nd time not long after calling the AWS Lambda Function the 1st time.  The issue is described in the [GitHub issue](https://github.com/johnt84/GitHubWebHookTool/issues/19) in my [GitHub repo](https://github.com/johnt84/GitHubWebHookTool).  

The source code is in the [GitHub Web Hook AWS Lambda Function](https://github.com/johnt84/GitHubWebHookTool/tree/master/GitHubWebHookAWSLambdaTool).  

Can you please help with this?

**Comment by ashishdhingra** [comment]
@johnt84 Based on what I see in your code, 
- The constructor for [HttpAPIClient](https://github.com/johnt84/GitHubWebHookTool/blob/ed5a415ee9565699470c0722b5f9de996093825d/GitHubWebHookEngine/API/HttpAPIClient.cs#L12) is accepting `HttpClient`. You code relies on call to [services.AddHttpClient()](https://github.com/johnt84/GitHubWebHookTool/blob/ed5a415ee9565699470c0722b5f9de996093825d/GitHubWebHookAWSLambdaTool/Startup.cs#L25). Looks like you are using [Typed Client](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/http-requests?view=aspnetcore-7.0#typed-clients), which is registered as **transient** with DI.
- However, the other services are registered as Scoped [here](https://github.com/johnt84/GitHubWebHookTool/blob/ed5a415ee9565699470c0722b5f9de996093825d/GitHubWebHookAWSLambdaTool/Startup.cs#L27), which uses are the same for a given request but differ across each new request.
- My guess is that the using transient HttpClient with scoped services is havign problem due to the similar issue reported https://github.com/Azure/azure-functions-host/issues/5098#issuecomment-548690627.

I'm unsure if this is the issue with the Lambda runtime environment. 

You might want to use the scoped `HttpAPIClient` using the workaround mentioned at https://stackoverflow.com/questions/49120741/register-scoped-based-httpclient-in-net-core-2. Please let me know if it works or if you were able to find solution yourself.

Thanks,
Ashish

**Reply by johnt84** [reply]
@ashishdhingra thanks for the quick response. Yeah it looks like the issue is occuring due to the project using a Typed HttpClient which is scoped but the other services are transient might be the issue.

I tried changing the other services to scoped and then tried using the solution from the workaround mentioned but above still getting the same issue.

Strangely I used a similar [solution](https://github.com/johnt84/GitHubWebHookTool/blob/master/GitHubWebHookAWSServerlessTool/Startup.cs) for an [AWS Serverless API solution](https://github.com/johnt84/GitHubWebHookTool/tree/master/GitHubWebHookAWSServerlessTool) and it works fine.

**Reply by ashishdhingra** [reply]
@johnt84 Thanks for your response. Were you able to figure out the differences?

**Reply by johnt84** [reply]
@ashishdhingra  No I can't see why the AWS Serveless API solution would work but the AWS Lambda app wouldn't work apart from that the AWS Serverless solution uses the LocalEntry point class which has the "startup" config code but the AWS Lambda app doesn't.

**Reply by johnt84** [reply]
@ashishdhingra  Do you think I should raise the Scoped HttpClient problem mentioned above as an issue with AWS Lambda?

**Reply by ashishdhingra** [reply]
@johnt84 I think this has something to do with the way Lambda function executes code in background.
CCing @normj for inputs.

**Reply by johnt84** [reply]
Great thanks @ashishdhingra

**Reply by johnt84** [reply]
@ashishdhingra Hi is there any update on this thanks?

**Reply by ashishdhingra** [reply]
CC @normj

**Reply by johnt84** [reply]
Thanks @ashishdhingra
